armi/reactor/converters/axialExpansionChanger/assemblyAxialLinkage.py

```diff
@@ -57,118 +69,191 @@ def _determineLinked(componentA, componentB):
+
+
+@dataclasses.dataclass
+class AxialLink(typing.Generic[Comp]):
+    """Small class for named references to objects above and below a specific object.
+
+    Axial expansion in ARMI works by identifying what objects occupy the same space axially.
+    For components, what components in the blocks above and below axially align? This is used
+    to determine what, if any, mass needs to be re-assigned across blocks during expansion.
+    For blocks, the linking determines what blocks need to move as a result of a specific block's
+    axial expansion.
+
+    Attributes
+    ----------
+    lower : Composite or None
+        Object below, if any.
+    upper : Composite or None
+        Object above, if any.
+
+    Notes
+    -----
+    This class is "templated" by the type of composite that could be assigned and fetched. A
+    block-to-block linkage could be type-hinted via ``AxialLink[Block]`` or ``AxialLink[Component]``
+    for component-to-component link.
+
+    See Also
+    --------
+    * :attr:`AxialAssemblyLinkage.linkedBlocks`
+    * :attr:`AxialAssemblyLinkage.linkedComponents`
+    """
+
+    lower: typing.Optional[Comp] = dataclasses.field(default=None)
+    upper: typing.Optional[Comp] = dataclasses.field(default=None)
+
+    def __getitem__(self, index: int) -> typing.Optional[Comp]:
+        """Get by position.
+
+        Discouraged since ``linkage.lower`` is more explicit and readable than ``linkage[0]``.
+
+        Parameters
+        ----------
+        index : int
+            ``0`` for :attr:`lower`, ``1`` for :attr:`upper`
+
+        Raises
+        ------
+        AttributeError
+            If ``index`` is not ``0`` nor ``1``
+        """
+        if index == 0:
+            return self.lower
+        if index == 1:
+            return self.upper
+        raise AttributeError(f"{index=}")
+
+    def __setitem__(self, index: int, o: Comp):
+        """Set by position.
+
+        Discouraged since ``linkage.upper = x`` is more explicit and readable than ``linkage[1] = x``.
+        """
+        if index == 0:
+            self.lower = o
+        elif index == 1:
+            self.upper = o
         else:
-            idA, odA = (
-                componentA.getCircleInnerDiameter(cold=True),
-                componentA.getBoundingCircleOuterDiameter(cold=True),
-            )
-            idB, odB = (
-                componentB.getCircleInnerDiameter(cold=True),
-                componentB.getBoundingCircleOuterDiameter(cold=True),
-            )
+            raise AttributeError(f"{index=}")
 
-            biggerID = max(idA, idB)
-            smallerOD = min(odA, odB)
-            if biggerID >= smallerOD:
-                # one object fits inside the other
-                linked = False
-            else:
-                linked = True
+    def __iter__(self):
+        return iter([self.lower, self.upper])
 
-    else:
-        linked = False
 
-    return linked
+if typing.TYPE_CHECKING:
```

Comment:
Can we move this up to just after the regular imports? It's strange to see importing this far down the file.

#wont-die-on-that-hill